File: main workshop/services/java/dynamic-pricing-service/src/main/java/com/powertoolsride/dynamicpricingservice/service/RushHourService.java
package com.powertoolsride.dynamicpricingservice.service;

import com.powertoolsride.dynamicpricingservice.model.SecretData;
import com.fasterxml.jackson.databind.ObjectMapper;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;

import java.math.BigDecimal;

public class RushHourService {
    private final SecretsManagerClient secretsManager;
    private final ObjectMapper objectMapper;
    private final String secretName;

    public RushHourService() {
        this.secretsManager = SecretsManagerClient.create();
        this.objectMapper = new ObjectMapper();
        this.secretName = System.getenv("RUSH_HOUR_MULTIPLIER_SECRET_NAME");
        if (secretName == null || secretName.isEmpty()) {
            throw new IllegalStateException("RUSH_HOUR_MULTIPLIER_SECRET_NAME environment variable is not set");
        }
    }

    public BigDecimal getRushHourMultiplier() {
        GetSecretValueRequest request = GetSecretValueRequest.builder()
            .secretId(secretName)
            .build();

        GetSecretValueResponse response = secretsManager.getSecretValue(request);
        
        SecretData secretData;
        try {
            secretData = objectMapper.readValue(response.secretString(), SecretData.class);
        } catch (Exception e) {
            throw new RuntimeException("Failed to deserialize secret data", e);
        }
        
        if (secretData == null) {
            throw new RuntimeException("Failed to deserialize secret data");
        }
        
        return secretData.rushHourMultiplier();
    }
}

File: main workshop/services/python/dynamic-pricing-service/rush_hour_service.py
import json
import os
from decimal import Decimal

import boto3

from models import SecretData


class RushHourMultiplierService:
    """Service for retrieving rush hour multiplier from AWS Secrets Manager"""

    def __init__(self):
        self.secrets_manager = boto3.client("secretsmanager")
        self.secret_name = os.environ.get("RUSH_HOUR_MULTIPLIER_SECRET_NAME", "")

        if not self.secret_name:
            raise ValueError(
                "RUSH_HOUR_MULTIPLIER_SECRET_NAME environment variable is not set"
            )

    def get_rush_hour_multiplier(self) -> Decimal:
        """Retrieve rush hour multiplier from AWS Secrets Manager"""
        response = self.secrets_manager.get_secret_value(SecretId=self.secret_name)

        if not response.get("SecretString"):
            raise Exception("Failed to deserialize secret data")

        secret_data_dict = json.loads(response["SecretString"])
        secret_data = SecretData(**secret_data_dict)

        return secret_data.rush_hour_multiplier
